### ISSUE TITLE
wrong type inferred on index access to tuple with variadic type parameter

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#4299]
**Describe the bug**

unpacked `TypeVarTuple` is counted as single element when inferring index-accessed tuple member.

**To Reproduce**

```python
from typing import Tuple, TypeVar
from typing_extensions import TypeVarTuple, Unpack, reveal_type

S = TypeVar("S")
T = TypeVar("T")
U = TypeVar("U")
V = TypeVar("V")
Ts = TypeVarTuple("Ts")
Us = TypeVarTuple("Us")

def a(w: Tuple[Unpack[Ts]]):
    reveal_type(w[0])  # *Ts, expected Union[*Ts]

def f(x: Tuple[S, T, Unpack[Ts], U, V]):
    reveal_type(x[0])  # S
    reveal_type(x[1])  # T
    reveal_type(x[2])  # *Ts, expected Union[*Ts, U]
    reveal_type(x[3])  # U, expected Union[*Ts, U, V]
    reveal_type(x[4])  # V, expected Union[*Ts, U, V]
    reveal_type(x[5])  # Union[S, T, *Ts, U, V] with error, expected Union[*Ts, U, V] without error
    reveal_type(x[-1])  # V
    reveal_type(x[-2])  # U
    reveal_type(x[-3])  # *Ts, expected Union[T, *Ts]
    reveal_type(x[-4])  # T, expected Union[S, T, *Ts]
    reveal_type(x[-5])  # S, expected Union[S, T, *Ts]
    reveal_type(x[-6])  # Union[S, T, *Ts, U, V] with error, expected Union[S, T, *Ts] without error

def g(x: Tuple[S, Unpack[Ts], T], y: Tuple[U, Unpack[Us], V]):
    reveal_type(x + y)  # Tuple[S, *Ts, T, U, *Us, V]
    reveal_type((x + y)[0])  # S
    reveal_type((x + y)[1])  # *Ts, expected Union[*Ts, T]
    reveal_type((x + y)[2])  # T, expected Union[*Ts, T, U]
    reveal_type((x + y)[3])  # U, expected Union[*Ts, T, U, *Us, V]
    reveal_type((x + y)[4])  # *Us, expected Union[*Ts, T, U, *Us, V]
    reveal_type((x + y)[5])  # V, expected Union[*Ts, T, U, *Us, V]
    reveal_type((x + y)[6])  # Union[S, *Ts, T, U, *Ts, V] with error, expected Union[*Ts, T, U, *Ts, V] without error
```
```
r.py:12:17 - information: Type of "w[0]" is "*Ts@a"

r.py:15:17 - information: Type of "x[0]" is "S@f"
r.py:16:17 - information: Type of "x[1]" is "T@f"
r.py:17:17 - information: Type of "x[2]" is "*Ts@f"
r.py:18:17 - information: Type of "x[3]" is "U@f"
r.py:19:17 - information: Type of "x[4]" is "V@f"
r.py:20:17 - information: Type of "x[5]" is "S@f | T@f | Union[*Ts@f] | U@f | V@f"
r.py:20:17 - error: Index 5 is out of range for type Tuple[S@f, T@f, *Ts@f, U@f, V@f] (reportGeneralTypeIssues)
r.py:21:17 - information: Type of "x[-1]" is "V@f"
r.py:22:17 - information: Type of "x[-2]" is "U@f"
r.py:23:17 - information: Type of "x[-3]" is "*Ts@f"
r.py:24:17 - information: Type of "x[-4]" is "T@f"
r.py:25:17 - information: Type of "x[-5]" is "S@f"
r.py:26:17 - information: Type of "x[-6]" is "S@f | T@f | Union[*Ts@f] | U@f | V@f"
r.py:26:17 - error: Index -6 is out of range for type Tuple[S@f, T@f, *Ts@f, U@f, V@f] (reportGeneralTypeIssues)

r.py:29:17 - information: Type of "x + y" is "tuple[S@g, *Ts@g, T@g, U@g, *Us@g, V@g]"
r.py:30:17 - information: Type of "(x + y)[0]" is "S@g"
r.py:31:17 - information: Type of "(x + y)[1]" is "*Ts@g"
r.py:32:17 - information: Type of "(x + y)[2]" is "T@g"
r.py:33:17 - information: Type of "(x + y)[3]" is "U@g"
r.py:34:17 - information: Type of "(x + y)[4]" is "*Us@g"
r.py:35:17 - information: Type of "(x + y)[5]" is "V@g"
r.py:36:17 - information: Type of "(x + y)[6]" is "S@g | Union[*Ts@g] | T@g | U@g | Union[*Us@g] | V@g"
r.py:36:17 - error: Index 6 is out of range for type tuple[S@g, *Ts@g, T@g, U@g, *Us@g, V@g] (reportGeneralTypeIssues)
```

**Expected behavior**

types inferred as commented above

**VS Code extension or command-line**

command-line tool, 1.1.283

**Comment by erictraut** [comment]
I agree there's a bug here, but I don't plan to implement the solution you've suggested here. 

The logic in pyright that handles index expressions for tuples is intended to handle only the case where the index is a literal integer expression and all elements in the tuple are of known type and length. The inclusion of an unpacked tuple violates the last part. That means that the special case logic should not apply, and the evaluated type should be the union of all types within the tuple.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.284, which I just published. It will also be included in a future release of pylance.
